### ISSUE TITLE
Deprecated components and "Lean Core" migration story

### ISSUE THREAD

**Issue by cknitt** [reasonml-community/reason-react-native#474]
How do we plan to handle deprecated components and the moving of components to community repos? (See "Lean Core", https://facebook.github.io/react-native/blog/2019/03/12/releasing-react-native-059#lean-core-is-underway and https://github.com/facebook/react-native/issues/23313.)

* Should we move deprecated components to a `ReactNative.Deprecated` namespace?
* How/where shall we provide bindings for the new components that were moved out of core?
   * in separate packages like `reason-react-native-webview` etc.?
   * within the `reason-react-native` package? If so, under what namespace?
* What is the migration story when a component is moved out from core? How do we avoid breaking people's code in that case?

@MoOx @sgny

**Comment by MoOx** [comment]
Here is how I see things

Part 1: use `[@deprecated]` on react component (`let make`) so people natively get a warning from reason/ocaml (and don't change anything beside this). The deprecation message will tell people to use a new package (see part 2)
Part 2: move deprecated compo as a `reason-react-native-*` package (we are in a monorepo, easy to handle technically speaking - we could imagine doing a `packages` folder at the root) and adjust (eg: webview as changed a little if I am correct)
Part 3: when things are totally removed from RN core, wait a few versions/months if that's not a big deal for us & then remove those from `ReactNative` namespace completely. People will have time to use new packages/bindings as they will have see annoying deprecation warnings for a while...

---

For part 2: we can start this way, and also accept various "well know" packages (eg: react-navigation bindings?) or we could make several repo. I would go monorepo but on the long run this might become  not that easy to maintain. We can start monorepo & see how things goes... (If we hit the "too many packages/bindings, I will see that's as a very good sign :) )

**Comment by cknitt** [comment]
Sounds great! 👍 I am all for monorepo, too! :-)

And what about module names?

As we have reason-react-native -> ReactNative, we would also have 
reason-react-native-webview -> ReactNativeWebView?

So the user would do

```re
open ReactNativeWebView;

[@react.component]
let make = () => <WebView ... />;
```

?

**Comment by MoOx** [comment]
Yeah I guess so. Easy to understand: reason-react-whatever => namespace is ReactWhatever.
Would avoid confusion for lots of people as soon as it's documented

**Comment by sgny** [comment]
I like your transition strategy and it would be great to have a single source for bindings.

Perhaps we can make use of [conditional compilation](https://bucklescript.github.io/docs/en/conditional-compilation) to return `deprecated` messages based on RN version (to be set by the user). Unfortunately that requires OCaml syntax at the moment and no timeframe for Reason syntax support as far as I know. For bindings we don't actually need JSX, so it should be relatively painless to use OCaml syntax.

**Comment by wokalski** [comment]
There's also a conditional compilation ppx that should work with Reason.
https://github.com/janestreet/ppx_optcomp

**Comment by sgny** [comment]
> There's also a conditional compilation ppx that should work with Reason.
> https://github.com/janestreet/ppx_optcomp

Thanks! Since so many features/props are added/replaced/taken out over RN versions, IMO, it would be nice to have some guidance from the compiler on what is available for some given RN version.

@MoOx @cknitt What do you think?

**Comment by MoOx** [comment]
As said in another thread, somewhere, I don't really want to got this way. My first fear is that as soon as you do that, you aren't able to reuse the same JS for multiple platform. This will be especially annoying during development if you use the RN js server while developing on both on iOS device & and Android one.
On top of that this will require from users to have things to configure to say "I am compilation to be for platform X"...
So, the way I see things would be more like:
- as soon as something is deprecated in RN, we add a `[@deprecated]` with a message about the version. Btw, if you are using RN X that's doesn't have feature A deprecated doesn't mean you aren't interested about the deprecation coming in version Y. On the contrary, I found this pretty nice to be able to get deprecation notice before they actually become annoying YELLOW WARNINGS.
- we will have to decide how we are going to follow RN version & our version anyway, even with conditional compilation. I found that's version grid are not optimal ([eg](https://github.com/react-native-community/react-native-svg#notice) as this is totally not intuitive. I was thinking about following RN version limited to X.Y omiting the Z, reserved to patch that can be from our side (eg: incorrect binding that needs a patch).
For example, this could mean reason-react-native 0.59.* would mean that it's fully compatible with RN 0.59.* (as said, bindings bug aside).

Regarding props/feature change:
- if a feature is added in RN, we can add it here asap. As soon as we document feature/props correctly (eg: on the website for each props: "available since vX.Y") it shouldn't be a problem. People cannot expect a feature "binded" to just work. We have to communicate on the fact that this project is just a safe binding, nothing more
- if a feature is removed, we can start with a deprecating notice a decide a strategy here. Whether we find something smart (eg: see above - not saying this idea is smart, but that's an idea), or we can just say "as soon as something is remove, you will have it in next minor version of the binding, but not in the one after". So if RN 0.60 remove something & our binding is at version 0.1.0, we can keep it for 0.2 & remove it in v0.3. Another way would be to keep deprecation notice for major: eg: RN 0.60 remove something, we are at 1.0.0, we keep it until 2.0.0 is out.

Btw, we should think about if we go to 0.x.0 (sorry if I am repeating myself) or if we want to go with x.0.0 (I have to say that I don't find useful the fact that lot's of project are using 0.x as soon as they are used widely... eg: RN, reason-react etc)...

Sorry for this huge response that is asking more questions than provide answers 🙃

**Comment by cknitt** [comment]
I like your suggestion about following RN version limited to X.Y omitting the Z.

But how could we have x.0.0 then? Do you mean you would like to have 59.x.y for RN 0.59.x?

**Comment by MoOx** [comment]
> But how could we have x.0.0 then? Do you mean you would like to have 59.x.y for RN 0.59.x?

We wouldn't haha. I guess if we choose to follow RN versions, we should stick to 0.X.*, * being our patches following X (same for RN, 0.X.Y, Y are RN patches, so bindings should still be valid) right?

**Comment by sgny** [comment]
> As said in another thread, somewhere, I don't really want to got this way. My first fear is that as soon as you do that, you aren't able to reuse the same JS for multiple platform. This will be especially annoying during development if you use the RN js server while developing on both on iOS device & and Android one.

My preference would indeed be to enforce `iOS`, `Android` and `Web` only options, but we could always allow users to disable conditional compilation via a designated parameter value. Then, debug builds could be platform independent. Since that's not the popular option, here I was simply suggesting that the deprecation message depend on RN version. I wouldn't want to sort through warnings which I do not have to address until I migrate. For example, I was going to propose we deprecate old accessibility API props in the -next bindings, but according to RN documentation

> accessibilityTraits will soon be deprecated. When possible, use accessibilityRole and accessibilityStates instead of accessibilityTraits and accessibilityComponentType

which suggests me that there is less than perfect feature parity with the new API, i.e. when it is not possible. I'm not exactly sure if that is indeed the case, but should users really get warnings because they chose to stick with an old version of RN for whatever reason? We could, also, have a parameter which could be set on or off to get such warnings about things to come. Switch them on, find out what part of your codebase may need to be refactored, turn it off and go back to things as usual.

> On top of that this will require from users to have things to configure to say "I am compilation to be for platform X"...
> So, the way I see things would be more like:
> 
> * as soon as something is deprecated in RN, we add a `[@deprecated]` with a message about the version. Btw, if you are using RN X that's doesn't have feature A deprecated doesn't mean you aren't interested about the deprecation coming in version Y. On the contrary, I found this pretty nice to be able to get deprecation notice before they actually become annoying YELLOW WARNINGS.
> * we will have to decide how we are going to follow RN version & our version anyway, even with conditional compilation. I found that's version grid are not optimal ([eg](https://github.com/react-native-community/react-native-svg#notice) as this is totally not intuitive. I was thinking about following RN version limited to X.Y omiting the Z, reserved to patch that can be from our side (eg: incorrect binding that needs a patch).
>   For example, this could mean reason-react-native 0.59.* would mean that it's fully compatible with RN 0.59.* (as said, bindings bug aside).
> 
> Regarding props/feature change:
> 
> * if a feature is added in RN, we can add it here asap. As soon as we document feature/props correctly (eg: on the website for each props: "available since vX.Y") it shouldn't be a problem. People cannot expect a feature "binded" to just work. We have to communicate on the fact that this project is just a safe binding, nothing more
> * if a feature is removed, we can start with a deprecating notice a decide a strategy here. Whether we find something smart (eg: see above - not saying this idea is smart, but that's an idea), or we can just say "as soon as something is remove, you will have it in next minor version of the binding, but not in the one after". So if RN 0.60 remove something & our binding is at version 0.1.0, we can keep it for 0.2 & remove it in v0.3. Another way would be to keep deprecation notice for major: eg: RN 0.60 remove something, we are at 1.0.0, we keep it until 2.0.0 is out.
> 
> Btw, we should think about if we go to 0.x.0 (sorry if I am repeating myself) or if we want to go with x.0.0 (I have to say that I don't find useful the fact that lot's of project are using 0.x as soon as they are used widely... eg: RN, reason-react etc)...
> 
> Sorry for this huge response that is asking more questions than provide answers 🙃

My concern is then we only support bug fixes in the latest version or have multiple branches to support. RN 0.60 discussion thread suggests major changes, it may be a while before some packages get updated to be compatible, etc. so we'd need to decide when we move over to 0.60. Perhaps this would be less of an issue, especially when we have almost perfect coverage of RN components and API. There's a few stuff I'm working on and will create PRs for but I notice very little missing and my concern may well be unfounded.

**Comment by wellmonge** [comment]
I'm able to help @MoOx

**Comment by MoOx** [comment]
@wellmonge  what do you have in mind?

**Comment by MoOx** [comment]
> We could, also, have a parameter which could be set on or off to get such warnings about things to come. Switch them on, find out what part of your codebase may need to be refactored, turn it off and go back to things as usual.

@sgny what can we do that easily? Is this possible currently without too much hassle?

> My concern is then we only support bug fixes in the latest version or have multiple branches to support. 

I clearly don't want to support multiple branch/version. I guess nobody does.

> I wouldn't want to sort through warnings which I do not have to address until I migrate

Why not? It's a warning, not an error.
Is there a way to "hide" warnings in reason? Can we provide a number or something so people can turn off some deprecated notice?

To make things simple I think we should try `@deprecate` as soon as something is deprecated in a stable version of RN.

Beside the warning thing (we could decide to not warn at all & just remove stuff when they are removed with a breaking change release), how do you think we should go for the version number?
Follow RN or not at all (at maybe maintain a grid somewhere...?)

Go for reason-react-native 0.1.0? Or 0.59.0?

**Comment by sgny** [comment]
> > We could, also, have a parameter which could be set on or off to get such warnings about things to come. Switch them on, find out what part of your codebase may need to be refactored, turn it off and go back to things as usual.
> 
> @sgny what can we do that easily? Is this possible currently without too much hassle?
> 

I haven't actually tried it yet, but I will see how much hassle that creates and report back.

> > My concern is then we only support bug fixes in the latest version or have multiple branches to support.
> 
> I clearly don't want to support multiple branch/version. I guess nobody does.
> 
> > I wouldn't want to sort through warnings which I do not have to address until I migrate
> 
> Why not? It's a warning, not an error.
> Is there a way to "hide" warnings in reason? Can we provide a number or something so people can turn off some deprecated notice?

`bsb` can be configured in `bsconfig.json` to ignore warnings with a given number. I don't know if `@deprecated` can throw such a numbered warning.

> 
> To make things simple I think we should try `@deprecate` as soon as something is deprecated in a stable version of RN.
> 
> Beside the warning thing (we could decide to not warn at all & just remove stuff when they are removed with a breaking change release), how do you think we should go for the version number?
> Follow RN or not at all (at maybe maintain a grid somewhere...?)
> 
> Go for reason-react-native 0.1.0? Or 0.59.0?

I'll have to think about this, but whatever versioning we choose, we could always have a grid stating which version of RN the bindings support.

**Comment by MoOx** [comment]
So should we got with 0.58, 0.59, 0.60 for incoming release? Or we could be lazy & start with 0.1.0 & revisit this problematic later. reason-react doesn't bother keeping in sync with react.

**Comment by MoOx** [comment]
Ooops :)

**Comment by cknitt** [comment]
I think following the RN version will be easiest for our users and help to minimize support efforts, as it makes it clear what the matching version is without having to read any documentation.

**Comment by MoOx** [comment]
For naming new packages, we could do

`react-native-something` => `@reason-react-native/something`

Exception made for react-navigation since this package doesn't follow the implicit `react-native-` namespace, so for this one we could do `reason-react-react-navigation`?

I have got the namespace `@reason-react-native` on npm :)

**Comment by MoOx** [comment]
I am working on netinfo bindings, will make a good example package in the repo :)

**Comment by sgny** [comment]
@MoOx I have already published bindings for `react-native-netinfo` as [reason-react-native-netinfo](https://www.npmjs.com/package/reason-react-native-netinfo). Are you working on those?

**Comment by MoOx** [comment]
Not working on those. That said right now I just made an empty skeleton for `@reason-react-native/*` packages so no big deal. I was thinking we could have in this repo at least the one under react-native-community org. Would you be ok with that (if so, we can get your work in here?)
About the bs namespace, I was thinking about `ReactNativeNetinfo` or `RNNetinfo` (because I think we should at least prefix each package with RN or something that indicate that's not for everything (even if package name is explicit).

**Comment by MoOx** [comment]
@sgny Would you be ok if I try to integrate your work in here as `@reason-react-native/netinfo`? Or would you prefer to keep them out of this repo?

**Comment by sgny** [comment]
@MoOx Sorry I couldn't reply earlier I was traveling and then had to catch up on my return.

I can contribute them here, should they go in a directory in the root of the project? Such as `reason-react-native-netinfo`?

**Comment by MoOx** [comment]
Np. No rush.

Can you put them in a folder named `@reason-react-native/netinfo` (so they match the package name, easier to discover)?

**Comment by sgny** [comment]
Sure, I may need a few days to check up on changes to `react-native-netinfo` and make sure all is fine.

**Comment by MoOx** [comment]
Seems up to date. We can still improve them later. I contributed to flow interface recently https://github.com/flow-typed/flow-typed/blob/master/definitions/npm/@react-native-community/netinfo_v4.x.x/flow_v0.69.x-/netinfo_v4.x.x.js if that can help

**Comment by sgny** [comment]
I looked quickly through the recent commits and I think there are no API changes.

I will release version 3.2.1 as the latest that does not require `jetifier` but following releases will require versions 4.x. Do you have any ideas on that?

**Comment by MoOx** [comment]
Doc should be enough about that don't you think?

**Comment by sgny** [comment]
The docs are quite clear, but I will simply make versions 4 the main release and not beta any longer.

**Comment by sgny** [comment]
@MoOx I think I have a reasonable solution; `@reason-react-native/netinfo` will depend on version 4. `reason-react-native-netinfo` will depend on version 3 but I will add a deprecation message pointing toward the new package.

**Comment by MoOx** [comment]
That's perfect!

**Comment by MoOx** [comment]
Next step is https://github.com/reasonml-community/reason-react-native/issues/594 !

**Comment by MoOx** [comment]
FYI, reason-react-native repo is now reason-react-native/react-native & will be published as `@reason-react-native/react-native` on npm.
We discussed about this with @cknitt, ideas was to have all package under `@reason-react-native` namespace on npm & have similar repo name. reason-react-navigation will become `@reason-react-native/navigation`

**Comment by sgny** [comment]
I commented on this on a `WebView` [issue](https://github.com/reason-react-native/webview/issues/1) but should we move any bindings for components removed from the core into `*-legacy` repos? So we'd have `webview` and `webview-legacy` and those still using RN 0.59 and below could keep using the RN bindings, but add whichever `legacy` bindings they require. This way, we could use `WebView` or `NetInfo` etc as the module names without conflicts.

Or perhaps just one additional legacy bindings repo where all such bindings go, since maintaining multiple EOL bindings would be an unnecessary burden.

**Comment by MoOx** [comment]
I wouldn't bother maintaining 2 bindings especially if removed from the core. Imo we should delete deprecated bindings in "major" (0.6x.y) releases.
As soon as we have things a bit more stable (repo/packages wise), I plan to work on 0.61 & delete unnecessary bindings. People will need to upgrade anyway...
We could still keep some 0.59(etc)-stable branches & push fixes if we feel the need (but I don't).

In short

>Or perhaps just one additional legacy bindings repo where all such bindings go, since maintaining multiple EOL bindings would be an unnecessary burden.

💯

**Comment by MoOx** [comment]
@cknitt @sgny I am having second thoughts on what I did for core & navigation package.
I think for making this easy we should have this packages:
- reason-react-native (core bindings), not as `@reason-react-native/reason-react-native` or `@reason-react-native/react-native` to make this longer that it should,
- reason-react-navigation, not `@reason-react-native/navigation` to avoid confusion with react-native-navigation, nor `@reason-react-native/react-navigation` because it's a bit long & react is redundant...
- have all others (community) packages like it's setup, as `@reason-react-native/something` for `react-native-something`.

You ok with this? I plan to cut a release for reason-react-navigation as 0.60.1 with all recent patches so we can move forward with 0.61

**Comment by sgny** [comment]
Package names are so inconsistent that we will have to compromise anyway. As in `react-native-webview` and `@react-native-community/checkbox`, located at `react-native-community/react-native-webview` and `react-native-community/react-native-checkbox`.

I think having everything under `@reason-react-native` has some benefits as one can clearly see which are the BS bindings. Accordingly, `@reason-react-native/react-native` is fine IMO. `reason-react-native` is also fine, TBH, but one would lose some of the organisational simplicity.

On the other hand, while `@reason-react-native/react-navigation` seems to have a somewhat redundant `react` in it, that can be considered a necessary evil to improve clarity. Still, that is clearly documented as bindings for `react-navigation` so I don't see a huge problem with the way it is now either.

Community packages are fine as they are now.

**Comment by sgny** [comment]
Should we move bindings for all removed components/APIs to a `something-legacy` repo?

I think naming modules as `WebView` instead of `ReactNativeWebview` to more closely mirror the JS side would be a win.
